tsl/src/continuous_aggs/invalidation.c

```diff
@@ -250,19 +250,178 @@ cut_invalidation_along_refresh_window(const CaggInvalidationState *state, int64
+	do                                                                                             \
+	{                                                                                              \
+		bool should_free;                                                                          \
+		HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);                    \
+		type form;                                                                                 \
+		form = (type) GETSTRUCT(tuple);                                                            \
+		(entry)->hyper_id = form->hypertable_id;                                                   \
+		(entry)->modification_time = form->modification_time;                                      \
+		(entry)->lowest_modified_value = form->lowest_modified_value;                              \
+		(entry)->greatest_modified_value = form->greatest_modified_value;                          \
+		(entry)->is_modified = false;                                                              \
+		ItemPointerCopy(&tuple->t_self, &(entry)->tid);                                            \
+                                                                                                   \
+		if (should_free)                                                                           \
+			heap_freetuple(tuple);                                                                 \
+	} while (0);
+
+void
+invalidation_entry_set_from_hyper_invalidation(Invalidation *entry, const TupleInfo *ti,
+											   int32 hyper_id)
+{
+	invalidation_entry_set(entry,
+						   ti,
+						   hypertable_id,
+						   Form_continuous_aggs_hypertable_invalidation_log);
+	/* Since hypertable invalidations are moved to the continuous aggregate
+	 * invalidation log, a different hypertable ID must be set (the ID of the
+	 * materialized hypertable). */
+	entry->hyper_id = hyper_id;
+}
+
+static void
+invalidation_entry_set_from_cagg_invalidation(Invalidation *entry, const TupleInfo *ti)
+{
+	invalidation_entry_set(entry,
+						   ti,
+						   materialization_id,
+						   Form_continuous_aggs_materialization_invalidation_log);
+}
+
+/*
+ * Try to merge two invalidations into one.
+ *
+ * Returns true if the invalidations were merged, otherwise false.
+ *
+ * Given that we scan ordered on lowest_modified_value, the previous and
+ * current invalidation can overlap in two ways (generalized):
+ *
+ * |------|
+ *    |++++++++|
+ *
+ * |-------------|
+ *    |++++++++|
+ *
+ * The closest non-overlapping case is:
+ *
+ * |--|
+ *    |++++++++|
+ *
+ */
+static bool
+invalidation_entry_try_merge(Invalidation *entry, const Invalidation *newentry)
+{
+	/* Quick exit if no overlap. We need to account for entries of size 1
+	 * (lowest = greatest). */
+	if (entry->lowest_modified_value < newentry->lowest_modified_value &&
+		entry->greatest_modified_value <= newentry->lowest_modified_value)
```

Comment:
It will not be sufficient when `entry->lowest_modified_value == entry->greatest_modified_value`. The only second condition will give `true`, while if both, it will be `false`.

Reply:
cc @mkindahl @erimatnor or do I miss something?

Reply:
And I guess there is no test for this case, when lowest and greatest of entry are equivalent, right? If so, it is good to have a test.

Reply:
I actually changed this to just

```
if (entry->greatest_modified_value < newentry->lowest_modified_value)
```

this means that an entry like [10,10] will overlap with another entry [10,10], like it should. I believe this is covered in tests.

One thing that makes this a bit confusing is that invalidations are, AFAIK, inclusive in both lowest and greatest, while, e.g., the refresh window is exclusive at the end (makes me think that maybe the refresh window should be inclusive in both ends too?).

Reply:
IMHO, having an upper end that is non-inclusive is a lot easier to work with. C++ has adopted this convention, as has a bunch of other languages for this reason.

Reply:
Inclusive on both ends seems to be better for implementation and works well for invalidation semantics. E.g., [10,10) is incorrect, and it should be [10, 10.1) or something like this depending on the type.

I don't remember that it was discussed how refresh window should be treated. I find no information in the docs if the current upper limit, which is `refresh-lag`, is inclusive or exclusive. Furthermore, the refresh window is transformed into bucket widths.